### PR TITLE
Add NEON assembly for rejection sampling

### PR DESCRIPTION
**Pull request by cothan**
[//]: # (SPDX-License-Identifier: CC-BY-4.0)

Improve the performance of Rejection Sampling. 

I should re-arrange the instructions, but I think when we have SLOTHY figures our the best instruction sequences, so I left it as future work. 

Performance improvement: average 8%.

### CODE REVIEW

**Comment by hanno-becker** [comment]
> I should re-arrange the instructions, but I think when we have SLOTHY figures our the best instruction sequences, so I left it as future work.

Yes, please write the code as clean as possible for now -- micro-optimization will be left to SLOTHY. (And if you have time, please feel free to try that out yourself)

File: mlkem/rej_uniform_asm.s
Comment (hanno-becker) [inline mlkem/rej_uniform_asm.s:32]: Don't you need some preamble/postamble for stack stash/restore?

File: mlkem/rej_uniform.c
Comment (hanno-becker) [inline mlkem/rej_uniform.c:289]: There's no need to do this here -- you can allocate stack as needed in the ASM?

Comment (cothan) [reply mlkem/rej_uniform.c:289]: You're right, I'm thinking of using `memcpy` to copy data from stack to `r`,  so I forgot to move the stack and memcpy to asm.

Comment (cothan) [reply mlkem/rej_uniform.c]: done in https://github.com/pq-code-package/mlkem-c-aarch64/pull/47/commits/3cf987be5dc19702a271ee94a7585fd202fde681

**Comment by cothan** [comment]
Okay, we will need to integrate arm runner to run this code in CI.

**Comment by hanno-becker** [comment]
@potsrevennil Didn't #44 add an Arm runner?

**Comment by hanno-becker** [comment]
@cothan I see CI for https://github.com/pq-code-package/mlkem-c-aarch64/pull/47/commits/9ee3e95a2486218faffb882940094f7b4e082033 failing because of linting, not because of an architecture problem.

**Comment by hanno-becker** [comment]
@cothan Can you mark the PR as DRAFT until you think it is ready to be merged, and clean up the history beforehand?

**Comment by cothan** [comment]
Sure!

**Comment by hanno-becker** [comment]
@cothan If you rebase on `main`, we should now be able to see separate CI jobs for macos-latest, pqcp-arm64, cbmc, and linting.

File: mlkem/rej_uniform_asm.s
Comment (hanno-becker) [inline mlkem/rej_uniform_asm.s:1]: @mkannwischer @cothan 

I'd suggest we have a separate directory `mlkem/asm` for all assembly, and perhaps further split into `mlkem/asm/clean` and `mlkem/asm/opt`, where `clean` are the handwritten versions and `opt` will be the micro-optimized ones once we have them.

WDYT

Comment (cothan) [reply mlkem/rej_uniform_asm.s:1]: I agree.

Comment (cothan) [reply mlkem/rej_uniform_asm.s:1]: Done in https://github.com/pq-code-package/mlkem-c-aarch64/pull/47/commits/f5d283fbe620b53af0e4d8528160b83ebd88404d

File: mlkem/rej_uniform_asm.s
Comment (cothan) [reply mlkem/rej_uniform_asm.s]: So I check, register from x8 to x17 can be used freely by callee as temporary register. Register x19-x28 must be reserved, I tried not to use x19-x28 so I think the current approach has no problem.

Comment (hanno-becker) [reply mlkem/rej_uniform_asm.s]: You might need some more registers when scheduling the code through SLOTHY, but we can add the necessary prologue/epilogue then.

For now, can you add a comment / warning please?

Comment (cothan) [reply mlkem/rej_uniform_asm.s]: Comment added in https://github.com/pq-code-package/mlkem-c-aarch64/pull/47/commits/f4ead9afa2b892eed146a027a4308c97e4771b9b. 
Thanks for the suggestion @hanno-becker !
I think at kernel level (if this work is adopted to kernel), people would use intra-procedure-call scratch registers `x16, x17`, thus I stick to strict ARM64 calling convention. `x8, x16, x17, x29, x30` registers are preserved.

File: mlkem/asm/clean/rej_uniform_asm.s
Comment (hanno-becker) [inline mlkem/asm/clean/rej_uniform_asm.s:35]: Why don't we just make it a precondition that `dst` is suitably aligned?

File: mlkem/asm/clean/rej_uniform_asm.s
Comment (hanno-becker) [inline mlkem/asm/clean/rej_uniform_asm.s:4]: I am skeptical that we should write this by hand -- compilers will utilize have a highly-optimized `memcpy` already. What's your motivation here?

File: mlkem/asm/clean/rej_uniform_asm.s
Comment (hanno-becker) [inline mlkem/asm/clean/rej_uniform_asm.s:188]: Can you leave a comment describing the usage and layout of the stack?

File: mlkem/asm/clean/rej_uniform_asm.s
Comment (hanno-becker) [inline mlkem/asm/clean/rej_uniform_asm.s:91]: This is not a secret division, is it? Can we not just use `div`?

File: mlkem/asm/clean/rej_uniform_asm.s
Comment (cothan) [reply mlkem/asm/clean/rej_uniform_asm.s:4]: I couldn't make it compile and link to external memcpy. 
I tried `extern memcpy`, still, the compile won't make it compile. I stuck for a while and decided to write my version. 
Since the data is small, ~600 bytes, an indepdent memcpy could save some code space. 
Is there a way to ask compiler `extern memcpy`?

File: mlkem/asm/clean/rej_uniform_asm.s
Comment (cothan) [reply mlkem/asm/clean/rej_uniform_asm.s:91]: Correct, it's not secret division, using `div` instruction is not optimized in this case, since we divide by constant. It's faster to multiple with inverse, isn't it?

File: mlkem/asm/clean/rej_uniform_asm.s
Comment (cothan) [reply mlkem/asm/clean/rej_uniform_asm.s:188]: Sure, will do now.

File: mlkem/asm/clean/rej_uniform_asm.s
Comment (hanno-becker) [reply mlkem/asm/clean/rej_uniform_asm.s:4]: I haven't yet read the core assembly below in detail but am surprised that we need a `memcpy` at all, actually.

Comment (cothan) [reply mlkem/asm/clean/rej_uniform_asm.s:4]: Honestly, I don't want to write my own version of `memcpy`. I spent so long and I can't figure out why `extern memcpy` or `.extern memcpy` doesn't compile. If I find a way to link `memcpy` I would definitely remove my own.

File: mlkem/asm/clean/rej_uniform_asm.s
Comment (cothan) [reply mlkem/asm/clean/rej_uniform_asm.s:197]: added in https://github.com/pq-code-package/mlkem-c-aarch64/pull/47/commits/908164c2534b4d6dc24e43cd1ad8cd6fe4ffcbd2

File: mlkem/asm/clean/rej_uniform_asm.s
Comment (hanno-becker) [reply mlkem/asm/clean/rej_uniform_asm.s:91]: I stand corrected, compilers also don't use `div 48` here.

Alternatively, though, one can just use `sub _, _, #48` in the loop, rather than having an explicit iteration counter. But SLOTHY would not yet support that, so let's leave it.

Comment (hanno-becker) [reply mlkem/asm/clean/rej_uniform_asm.s:91]: Compilers agree with you that the approach above is faster. Still, this is a one-off, so I don't think this is worth optimizing, given that it comes with some cognitive and verification overhead?

Comment (cothan) [reply mlkem/asm/clean/rej_uniform_asm.s:91]: Okay, I'll use `div` instruction directly. Update soon.

Comment (cothan) [reply mlkem/asm/clean/rej_uniform_asm.s]: done in https://github.com/pq-code-package/mlkem-c-aarch64/pull/47/commits/b42c7a17e06fd23922bdb6a3ac518cfe517a6e59

File: mlkem/asm/clean/rej_uniform_asm.s
Comment (cothan) [reply mlkem/asm/clean/rej_uniform_asm.s]: okay, I figure it out, turn out on macos the symbol for `memcpy` is `_memcpy`, and linux is `memcpy`. 
I added a macro for to switch. Now it's worked! the `simd_memcpy` is removed. 
Commit history is cleaned.

**Comment by cothan** [comment]
Okay, after I removed the `simd_memcpy` and replace with `memcpy`. It works on Mac. 
I can't wrap my head why the `arm64` version land on the `#ifdef __APPLE_` branch and take on `_memcpy` while on Linux. 
@.@ I need to investigate why the compiler on `arm64` land on `__APPLE__` branch.

**Comment by cothan** [comment]
Okay, I know why. The assembly extension must be capital. `.S` instead of `.s`. :(( https://stackoverflow.com/a/30327226

**Comment by cothan** [comment]
My M1 benchmark:

**Before:**

| Parameter           | MLKEM512 | MLKEM768 | MLKEM1024 |
|---------------------|----------|----------|-----------|
| keypair cycles      | 60678    | 103056   | 161138    |
| encaps cycles       | 75188    | 122977   | 184479    |
| decaps cycles       | 93538    | 146287   | 214105    |

**After:**

| Parameter           | MLKEM512 | MLKEM768 | MLKEM1024 |
|---------------------|----------|----------|-----------|
| keypair cycles      | 59284    | 99135    | 153797    |
| encaps cycles       | 73790    | 118616   | 176734    |
| decaps cycles       | 91968    | 141692   | 206045    |

**Comment by cothan** [comment]
The number is much bigger (probably same ratio) compare on RPi 4. 

This branch: 
https://github.com/pq-code-package/mlkem-c-aarch64/actions/runs/9718493064?pr=47

Main branch: 
https://github.com/pq-code-package/mlkem-c-aarch64/actions/runs/9714693885

**Review by hanno-becker — CHANGES_REQUESTED** [review]
@cothan @mkannwischer I would like all assembly in MLKEM-C-AArch64 be guarded by at least one compile-time flag, so we can easily switch it off, for example to test whether functional issues are related to ASM or not.

@cothan Could you consider where such a flag should live, introduce it, and guard the newly introduced ASM of this PR accordingly?

**Comment by hanno-becker** [comment]
Apologies @cothan, I've just discussed this with @mkannwischer again. We believe that this work is low-priority and would rather slow us down on the path towards an initial release, seeing that we strive for functional verification of all assembly. We therefore feel that the PR should be put on hold until after the first release.

I'm sorry @cothan that we did not sort this out earlier.

**Comment by cothan** [comment]
No problem. Thanks for letting me know.

**Comment by mkannwischer** [comment]
Turning this into a draft for now.

**Comment by hanno-becker** [comment]
@cothan I feel that now would be a good time to revive and update this pull request, rejection sampling being the last performance-relevant function that is not yet written in assembly. I don't know if/when we will manage to verify it, but if not, we may hide it behind an appropriate flag if needed.

Do you have time to work on this?

**Comment by cothan** [comment]
@hanno-becker , yup, I have time to work on this PR, I would like to use this PR to try out SLOTHY as well.

**Comment by hanno-becker** [comment]
@cothan Nice! Let's start with the clean version in this PR, and do an optimized one as a follow-up. You will need to add quite a few instructions to SLOTHY.

**Pull request closed without merging**
